### PR TITLE
Thread transform

### PR DESCRIPTION
**Pull request by mereacre**
- Integrated all @aloisklink pull requests
 - Translated from running processes to running threads
 - Loads of refactoring

@aloisklink could you please check if all of it compiles on owrt and maybe if you find some other optimizations in cmake that we need to do?

### CODE REVIEW

**Review by aloisklink — COMMENTED** [review]
Hi @mereacre, I'm probably going to need to complete this review tomorrow.

I'll submit my current review, but I'm still only at the `src/capture/cleaner_middleware/cleaner_middleware.c` function, so still a while to go.

Also, I'm trying to use https://conventionalcomments.org/ so you know that **todo:** are the actual important review comments, everything else is less and less important. Let me know if it helps.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -1,45 +1,41 @@
...
-  target_link_libraries(engine PRIVATE crypt_service net log iface_mapper os sqlite_macconn_writer sqlite_fingerprint_writer firewall_service eloop supervisor network_commands radius_service mac_mapper ap_service firewall_service dhcp_service)
-else ()
-  target_link_libraries(engine PRIVATE net log iface_mapper os sqlite_macconn_writer sqlite_fingerprint_writer firewall_service eloop supervisor network_commands radius_service mac_mapper ap_service firewall_service dhcp_service)
+  target_link_libraries(engine PRIVATE crypt_service)
+endif ()
+if (USE_MDNS_SERVICE)
+  target_compile_definitions(engine PUBLIC WITH_MDNS_SERVICE)
+  target_link_libraries(engine PRIVATE mdns_service)
 endif ()
+target_link_libraries(engine PRIVATE capture_service net log iface_mapper os sqlite_macconn_writer firewall_service eloop supervisor network_commands mac_mapper ap_service firewall_service dhcp_service)
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/CMakeLists.txt]: You are right. We should only link if USE_CAPTURE_SERVICE is defined.

File: src/capture/cleaner_middleware/cleaner_middleware.c [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+#include "middlewares.h"
+
+#include "../../utils/eloop.h"
+#include "../../utils/utarray.h"
+
+#define CLEANER_PROCESS_INTERVAL                                               \
+  5 /* Frequency in sec to run the cleaner function*/
+#define CLEANER_GROUP_INTERVAL                                                 \
+  1000 /* Number of rows to sum from the pcap metadata to calculate the store  \
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/capture/cleaner_middleware/cleaner_middleware.c:47]: Good point. I still need to test this code , though.

File: src/capture/cleaner_middleware/cleaner_middleware.c [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+static const UT_icd pcap_file_meta_icd = {sizeof(struct pcap_file_meta), NULL,
+                                          NULL, NULL};
+
+struct cleaner_middleware_context {
+  char pcap_path[MAX_OS_PATH_LEN];
+  uint64_t store_size;
+  uint64_t low_timestamp;
+  uint64_t store_sum;
+  uint64_t next_timestamp;
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/capture/cleaner_middleware/cleaner_middleware.c:60]: Could you please write as a TODO or issue? I will need to change in many places.

File: src/capture/cleaner_middleware/cleaner_middleware.c [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+                            pcap_meta_arr) < 0) {
+      log_error("get_pcap_array fail");
+      utarray_free(pcap_meta_arr);
+      return -1;
+    }
+
+    while ((p = (struct pcap_file_meta *)utarray_next(pcap_meta_arr, p)) !=
+           NULL) {
+      if ((path = construct_path(cleaner_context->pcap_path, p->name)) ==
+          NULL) {
```
[the remark this answers is not in the document]
Comment (mereacre) [reply src/capture/cleaner_middleware/cleaner_middleware.c:86]: I always thought that I might use the version before C99. Quite naive I'm.

File: src/capture/cleaner_middleware/cleaner_middleware.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (aloisklink) [reply src/capture/cleaner_middleware/cleaner_middleware.c:60]: Issue added in https://github.com/nqminds/EDGESec/issues/158!

**Pull request merged**